### ISSUE TITLE
Glyph Request: The Control Pictures block

### ISSUE THREAD

**Issue by Ryan1729** [tonsky/FiraCode#764]
The Unicode [Control Pictures block](https://en.wikipedia.org/wiki/Control_Pictures) contains glyphs for displaying otherwise unprintable control characters from ASCII, such as tab or line feed. It is sometimes useful to be able to see which, if any of these characters are contained in a file. If a font has these Control Pictures glyphs then a viewing program can display them instead of those characters. For the first 33 characters this turns out to be a simple addition of 0x2400 to each character below 33.

The most common case where it is useful to see which control characters are present is when a file has been edited with a program which has changed the line endings from windows-style CRLF to unix-style LF or vice-versa. So the ␍ and ␊ characters would probably see the most use. However, in whtespace sensitive languages like Python it can also be useful to see exactly how many spaces and or tabs are on each line of a file.

**Comment by tonsky** [comment]
Thanks, a great suggestion! Do you know any programs that can display those?

**Comment by Ryan1729** [comment]
Sorry for not responding here. I ended up editing a copy of the font to stuff glyphs for CR and LF into it, which solved my immediate problem, and it hasn't really been an issue for me since.

**Comment by Ryan1729** [comment]
Although now that I think about it, having all of these available will make debugging something I plan to do later easier. Anyway, thank you for adding these!

**Comment by Quantlyx** [comment]
> Although now that I think about it, having all of these available will make debugging something I plan to do later easier. Anyway, thank you for adding these!

Have these been added? I still can't get them to show in Putty...

**Comment by Ryan1729** [comment]
> Have these been added? I still can't get them to show in Putty...

Yes they have. Here's the table from the relevant [Wikipedia page](https://en.wikipedia.org/wiki/Control_Pictures) displayed using Fira Code in [a program I wrote](https://github.com/Ryan1729/rote):

![2022-08-08_644x128](https://user-images.githubusercontent.com/2133026/183535591-9e3de4fa-7e04-4d8a-9503-6824b8b9b153.png)

And here's a copy of the text shown in that screenshot
```
U+240x ␀ ␁ ␂ ␃ ␄ ␅ ␆ ␇ ␈ ␉ ␊ ␋ ␌ ␍ ␎ ␏
U+241x ␐ ␑ ␒ ␓ ␔ ␕ ␖ ␗ ␘ ␙ ␚ ␛ ␜ ␝ ␞ ␟
U+242x ␠ ␡ ␢ ␣ ␤ ␥ ␦ 
```

Mayne confirm you are using [the currently newest release](https://github.com/tonsky/FiraCode/releases/tag/6.2)?

**Comment by Quantlyx** [comment]
First of all: Thanks a lot!
So it turns out im neither missing U+000D nor U+240D or at least I don't need them. There is definitely a bug which causes CR and such symbols to be displayed in kinda comic sans at low font-sizes.

That said I need the CR-Arrow (aka. ↵).
That said [U+21B0 - U+21B3] seem to exist but neither U+21B4 nor my desired U+21B5 are to be found.

If you cold double-check this for me as I do not have any decent knowlege of font mechanics.

You have stated that you build your own custom font out of FiraCode to add CR on when it was not implemented yet. Maybe you could recommend an easy to use software just to add or edit some characters? 
Thanks a lot in advance!

**Comment by Quantlyx** [comment]
Maybe @tonsky could take a look at that.
(Missing: U+21B4 and U+21B5)

**Comment by Ryan1729** [comment]
> You have stated that you build your own custom font out of FiraCode to add CR on when it was not implemented yet. 

I had attempted that, and I think I had managed to edit the font, but if memory serves, it turned out that the font rendering software library I was using seemed to have a special case built-in such that it skipped even looking for a character for U+000D. So [I ended up mapping control characters to their control pictures equivalents myself](https://github.com/Ryan1729/rote/blob/d0799193e795ad4f6058b0153dca70610e8058be/libs/platform_layer/libs/window_layer/gl_layer/text_rendering/src/text_rendering.rs#L773-L781). I suspect that you unfortunately, don't have that option for your use case.

> Maybe you could recommend an easy to use software just to add or edit some characters?

I recall trying to use [Font Forge](https://fontforge.org/en-US/), but I think I ended up using [Birdfont](https://birdfont.org/) in the end. I also unfortunately recall both of those programs having a bit of a learning curve, and most documentation not being focused on the "copy this character to this slot" use case. Honestly I'd be interested if someone was able to find or create some easy to use software for this purpose.
